nvflare/app_opt/xgboost/controller.py

```diff
@@ -0,0 +1,145 @@
+from nvflare.apis.impl.controller import Controller
+from nvflare.apis.shareable import Shareable
+from nvflare.apis.signal import Signal
+from nvflare.apis.utils.common_utils import get_open_ports
+from nvflare.apis.workspace import Workspace
+from nvflare.app_common.app_constant import AppConstants
+
+from .constants import XGBShareableHeader
+
+
+class XGBFedController(Controller):
+    def __init__(self, train_timeout: int = 300, port: int = None):
+        """XGBoost federated training controller
+
+        Args:
+            train_timeout (int, optional): Time to wait for clients to do local training in seconds.
+            port (int, optional): the port to open XGBoost FL server
+
+        Raises:
+            TypeError: when any of input arguments does not have correct type
+            ValueError: when any of input arguments is out of range
+        """
+        super().__init__()
+
+        if not isinstance(train_timeout, int):
+            raise TypeError("train_timeout must be int but got {}".format(type(train_timeout)))
+
+        self._port = port
+        self._xgb_fl_server = None
+        self._participate_clients = None
+        self._rank_map = None
+        self._secure = False
+        self._train_timeout = train_timeout
+
+    def _get_certificates(self, fl_ctx: FLContext):
+        workspace: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        bin_folder = workspace.get_startup_kit_dir()
+        server_cert_path = os.path.join(bin_folder, "server.crt")
+        if not os.path.exists(server_cert_path):
+            self.log_error(fl_ctx, "Missing server certificate (server.crt)")
+            return False
+        server_key_path = os.path.join(bin_folder, "server.key")
+        if not os.path.exists(server_key_path):
+            self.log_error(fl_ctx, "Missing server key (server.key)")
+            return False
+        client_cert_path = os.path.join(bin_folder, "rootCA.pem")
+        if not os.path.exists(client_cert_path):
+            self.log_error(fl_ctx, "Missing client certificate (rootCA.pem)")
+            return False
+        self._server_cert_path = server_cert_path
+        self._server_key_path = server_key_path
+        self._client_cert_path = client_cert_path
+
+    def start_controller(self, fl_ctx: FLContext):
+        self.log_info(fl_ctx, f"Initializing {self.__class__.__name__} workflow.")
+        # Assumption: all clients are used
+        clients = self._engine.get_clients()
+        # Sort by client name so rank is consistent
+        clients.sort(key=lambda client: client.name)
+        rank_map = {clients[i].name: i for i in range(0, len(clients))}
+        self._rank_map = rank_map
+        self._participate_clients = clients
+
+        if not self._port:
+            self._port = get_open_ports(1)[0]
+
+        self.log_info(fl_ctx, f"Starting XGBoost FL server on port {self._port}")
+
+        if not self._get_certificates(fl_ctx):
```

Comment:
This should be in sync with NVFlare mode. If NVFlare is in secure mode, certificate should be required.

Reply:
thanks, I've fixed it.